src/compiler/lowering/MatMulOpLowering.cpp

```diff
@@ -14,198 +14,564 @@
+          "Vector size bits must be larger or equal to 0.");
+    vec_size_bits = s;
+    return *this;
+  }
+  LowerMatMulOpOptions &enableTiling(bool b = true) {
+    tile = b;
+    return *this;
+  }
+  int getVecSize(int bitwidth) const {
+    if (vec_size_bits > 0) {
+      return std::max(1, vec_size_bits / bitwidth);
+    } else {
+      return 1;
     }
+  }
+};
 
-    // col loop body
-    rewriter.setInsertionPointToStart(colLoop.getBody());
+llvm::SmallVector<AffineForOp, 3>
+affineMatMul(mlir::Value &lhs, mlir::Value &rhs, mlir::Value &output,
+             ConversionPatternRewriter &rewriter, mlir::Location loc,
+             ArrayRef<int64_t> lhsShape, ArrayRef<int64_t> rhsShape,
+             mlir::MLIRContext *ctx, SmallVector<AffineForOp, 3> &loops) {
+  // row loop
+  auto rowLoop = rewriter.create<AffineForOp>(loc, 0, lhsShape[ROW], 1);
+  // row loop body
+  rewriter.setInsertionPointToStart(rowLoop.getBody());
+  // col loop
+  auto colLoop = rewriter.create<AffineForOp>(loc, 0, rhsShape[COL], 1);
+  // col loop body
+  rewriter.setInsertionPointToStart(colLoop.getBody());
+  // fma loop
+  auto fmaLoop = rewriter.create<AffineForOp>(loc, 0, rhsShape[ROW], 1);
+  // inner loop body
+  rewriter.setInsertionPointToStart(fmaLoop.getBody());
+
+  auto a = rewriter.create<AffineLoadOp>(
+      loc, lhs,
+      ValueRange{rowLoop.getInductionVar(), fmaLoop.getInductionVar()});
+  auto b = rewriter.create<AffineLoadOp>(
+      loc, rhs,
+      ValueRange{fmaLoop.getInductionVar(), colLoop.getInductionVar()});
+  auto c = rewriter.create<AffineLoadOp>(
+      loc, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+  // split out into add and multiply
+  Value res = rewriter.create<LLVM::FMAOp>(loc, a, b, c);
+
+  rewriter.create<AffineStoreOp>(
+      loc, res, output,
+      ValueRange{rowLoop.getInductionVar(), colLoop.getInductionVar()});
+
+  // AffineYieldOp at end of loop blocks
+  rewriter.setInsertionPointAfter(fmaLoop);
+  rewriter.setInsertionPointAfter(colLoop);
+  rewriter.setInsertionPointAfter(rowLoop);
+
+  loops.push_back(rowLoop);
+  loops.push_back(colLoop);
+  loops.push_back(fmaLoop);
+  return loops;
+}
 
-    loopIvs.push_back(rowLoop.getInductionVar());
-    loopIvs.push_back(colLoop.getInductionVar());
-    loopIvs.push_back(innerLoop.getInductionVar());
+llvm::SmallVector<AffineForOp, 3>
+vectorizedAffineMatMul(mlir::Value &lhs, mlir::Value &rhs, mlir::Value &output,
+                       ConversionPatternRewriter &rewriter, mlir::Location loc,
+                       ArrayRef<int64_t> lhsShape, ArrayRef<int64_t> rhsShape,
+                       mlir::MLIRContext *ctx,
+                       llvm::SmallVector<AffineForOp, 3> &loops,
+                       Type elementType, int64_t vec_size) {
+  auto vec_Type = mlir::VectorType::get({vec_size}, elementType);
+
+  // TODO: We need an option to enable smaller vector sizes for the ends of each
```

Comment:
TODO, in case this is out of scope for this PR, maybe it is wort it to create an issue with more explanation and  reasoning for the suggested change and reference the issue here as `TODO(#xxx)`

Reply:
I will create a PR for this.